client/TracyProfiler.cpp

```diff
@@ -2907,8 +2915,195 @@ void Profiler::HandleSymbolQuery( uint64_t symbol )
+    // (It's an abuse of MappingInfo since this address range is not necessarily
+    // one mapping).
+    MappingInfo address_range;
+    address_range.start_address = start_address;
+    address_range.end_address = end_address;
+    // Comparison function for std::lower_bound. Returns true if all addresses in `m1`
+    // are lower than all addresses in `m2`.
+    auto Compare = []( const MappingInfo& m1, const MappingInfo& m2 ) {
+        // '<=' because the address ranges are half-open intervals, [start, end).
+        return m1.end_address <= m2.start_address;
+    };
+    MappingInfoIterator it = std::lower_bound( mappings.begin(), mappings.end(), address_range, Compare );
+    *start_iterator = it;
+    std::uintptr_t remainder_start_address = start_address;
+    // Loop over all contiguous mappings intersecting the requested address range.
+    while( it != mappings.end() && it->start_address < end_address ) {
+        if (it->start_address > remainder_start_address) {
+            // The current mapping starts after the remainder of the address
+            // range that we needs to cover, so some of that address range
+            // isn't covered by any mapping.
+            return false;
+        }
+        remainder_start_address = it->end_address;
+        ++it;
+    }
+    // Some of the address range isn't covered by any mapping.
+    if (remainder_start_address < end_address) return false;
+    *end_iterator = it;
+    if (*end_iterator < *start_iterator) {
+        abort();
+    }
+    if (*end_iterator > (*start_iterator) + 1) {
+        abort();
+    }
+    return true;
+}
+                      
+
+// Attempts to make the specified `mapping` readable if it isn't already.
+// Returns true if and only if the mapping is readable.
+bool EnsureReadable( MappingInfo* mapping )
+{
+    if( mapping->perm_r )
+    {
+        // The mapping is already readable.
+        return true;
+    }
+    if( mapping->mprotect_fails )
+    {
+        // We have already see mprotect failing on this mapping.
+        return false;
+    }
+    int prot = PROT_READ;
+    if( mapping->perm_w ) prot |= PROT_WRITE;
+    if( mapping->perm_x ) prot |= PROT_EXEC;
+    if( mprotect( reinterpret_cast<void*>( mapping->start_address ),
+                 mapping->end_address - mapping->start_address, prot ) == -1 )
+    {
+        // Failed to make the mapping readable. Update `mapping`
+        // so the next call will be fast.
+        mapping->mprotect_fails = true;
+        return false;
+    }
+    // The mapping is now readable. Update `mapping` so the next call will be fast.
+    mapping->perm_r = true;
+    return true;
+}
+
+// Attempts to make the specified address range readable.
+bool EnsureReadable( std::uintptr_t start_address, std::uintptr_t end_address )
+{
+    // Lock ensuring reentrancy of this function.
+    static std::mutex s_mutex;
+    const std::lock_guard<std::mutex> lock( s_mutex );
```

Comment:
This lock is not needed.

Reply:
done (wasn't sure if more than 1 thread might call this, and there was some mutex stuff in this file).